x/dsmr/node_test.go

```diff
@@ -994,6 +1022,287 @@ func TestNode_NewBlock_IncludesChunkCerts(t *testing.T) {
 	}
 }
 
+// TestDuplicateChunksElimination tests that duplicate chunks that have appeared before are getting correctly eliminated.
+func TestDuplicateChunksElimination(t *testing.T) {
+	r := require.New(t)
+
+	node := newTestNode(t)
+
+	blk := Block{
+		ParentID: ids.GenerateTestID(),
+		Hght:     1,
+		Tmstmp:   1,
+		blkID:    ids.GenerateTestID(),
+	}
+	r.NoError(node.Accept(context.Background(), blk))
+
+	chunk, chunkCert, err := node.BuildChunk(
+		context.Background(),
+		[]tx{
+			{
+				ID:     ids.GenerateTestID(),
+				Expiry: 4,
+			},
+		},
+		4,
+		codec.Address{},
+	)
+	r.NoError(err)
+
+	blk = Block{
+		ParentID: ids.GenerateTestID(),
+		Hght:     2,
+		Tmstmp:   2,
+		blkID:    ids.GenerateTestID(),
+		ChunkCerts: []*ChunkCertificate{
+			&chunkCert,
+		},
+	}
+	r.NoError(node.Accept(context.Background(), blk))
+
+	r.NoError(node.storage.AddLocalChunkWithCert(chunk, &chunkCert))
+	_, err = node.BuildBlock(context.Background(), blk, 3)
+	r.ErrorIs(err, ErrAllChunkCertsDuplicate)
+
+	// make sure that it's not the case with any other chunk.
+	anotherChunk, anotherChunkCert, err := node.BuildChunk(
+		context.Background(),
+		[]tx{
+			{
+				ID:     ids.GenerateTestID(),
+				Expiry: 4,
+			},
+		},
+		4,
+		codec.Address{},
+	)
+	r.NoError(err)
+	r.NoError(node.Accept(context.Background(), blk))
+
+	r.NoError(node.storage.AddLocalChunkWithCert(anotherChunk, &anotherChunkCert))
+	_, err = node.BuildBlock(context.Background(), blk, 3)
+	r.NoError(err)
+}
+
+func TestNode_Verify_Chunks(t *testing.T) {
+	r := require.New(t)
+	networkID := uint32(123)
+	chainID := ids.Empty
+	sk1, err := bls.NewSecretKey()
+	r.NoError(err)
+	pk := bls.PublicFromSecretKey(sk1)
+	signer := warp.NewSigner(sk1, networkID, chainID)
+	r.NoError(err)
+
+	initChunks := func(node *Node[tx]) ([]Chunk[tx], []*ChunkCertificate) {
+		var chunks []Chunk[tx]
+		var chunksCerts []*ChunkCertificate
+		for expiry := int64(0); expiry < 5; expiry++ {
+			chunk, cert, err := node.BuildChunk(
+				context.Background(),
+				[]tx{
+					{
+						ID:     ids.GenerateTestID(),
+						Expiry: expiry,
+					},
+				},
+				expiry,
+				codec.Address{},
+			)
+			r.NoError(err)
+			chunks = append(chunks, chunk)
+			chunksCerts = append(chunksCerts, &cert)
+		}
+		return chunks, chunksCerts
+	}
```

Comment:
Can we just move this to the test body? It's only used once so I find that this makes the test more confusing to read.

Reply:
sure.